### PR TITLE
fix: guard float → int casts against NaN / ±inf / overflow (#1232)

### LINKED ISSUE #1232: bug: fptosi-based narrowing is UB for NaN / inf / out-of-range inputs
## Context

Surfaced by CodeRabbit on PR #1229 (#1192 implementation review):
https://github.com/t0k0sh1/ry/pull/1229#discussion_r3106732047

#1192 added two implicit `f64 → i64` coercion branches:
- `coerceToLowLevelType` (`src/codegen_call_user.cpp`) — reaches var decl / reassign
- `applyCompoundOp` (`src/codegen_stmt_misc.cpp`) — reaches compound assign

Both use `CreateFPToSI`. Per LLVM LangRef, `fptosi` on NaN, ±inf, or magnitudes outside the target integer range produces **poison**, which propagates through downstream computations as undefined behavior.

The **pre-existing** explicit cast `x as int` (`src/codegen_expr_cast.cpp`) has the same UB — #1192 only widened the number of paths that can hit it.

## Reproduce

```ry
# Implicit narrowing (new in #1192)
x: int = 1.0 / 0.0     # produces +inf → fptosi(inf, i64) = poison
print(x)               # arbitrary value, no diagnostic

x: int = 5
x /= 0                 # +inf → poison via applyCompoundOp
print(x)               # arbitrary value

# Pre-existing explicit-cast path
y = (1.0 / 0.0) as int # same UB
print(y)
```

## Expected

Either:
1. **Runtime trap**: emit a branch that compares against `numeric_limits<int64_t>::max/min` and `CreateFCmpORD` for NaN, branching to a runtime error with a clear message (e.g. `"cannot convert NaN to int"`, `"float value out of int range"`).
2. **Saturation**: use `llvm.fptosi.sat.*` intrinsics (LLVM 20+) which define behavior as saturation to min/max for OOB and zero for NaN.

## Scope

One unified fix should cover all fptosi sites:
- `src/codegen_expr_cast.cpp` (`as int` cast, pre-existing)
- `src/codegen_call_user.cpp:714` (coerceToLowLevelType f64→i64, #1192)
- `src/codegen_stmt_misc.cpp:86` (applyCompoundOp f64→i64, #1192)

Also audit the reverse direction (`i64 → f64` via `CreateSIToFP`) — that path is total (all i64 values have a representable f64, though some lose precision), so no UB, but worth documenting in KNOWLEDGE.md.

## Notes

#1192's plan explicitly deferred this to a follow-up to avoid scope creep (see plan "Risks / Follow-up" section). KNOWLEDGE.md entry for #1192 already notes the UB caveat; this issue tracks the proper fix.

## Decision needed

- Trap vs saturation? Saturation is deterministic and closer to "do what users expect"; trap is safer for catching bugs early.
- Same behavior for `as int` (pre-existing) and implicit coercion (#1192)?

### PR DESCRIPTION
**Pull request by t0k0sh1**
## Summary

- Introduce `CodeGen::emitCheckedFPToInt` — a unified helper that guards every user-visible float → integer narrowing with a NaN / ±inf / out-of-range check. On failure the program prints `runtime error: cannot convert <value> to <type>` to stderr and exits with status `1`, matching Ry's existing policy for overflow / division-by-zero / bounds checks.
- Replace all 12 direct `CreateFPToSI` / `CreateFPToUI` call sites (9 `as` casts in `codegen_expr_cast.cpp`, the implicit `float → int` coercion in `coerceToLowLevelType`, the compound-assign f64 → i64 path in `codegen_stmt_misc.cpp`, and the math rounding path in `codegen_call.cpp`) with the helper.
- Fix the INT64_MIN boundary bug in the math path: the previous `fabs(x) >= 2^63` overflow guard spuriously rejected `-2^63` (exactly representable in f64) and missed cases where the **result** rounded out of range (e.g. `ceil(9.22e+18)` past `INT64_MAX`). The new helper uses half-open intervals `[-2^(W-1), 2^(W-1))` and runs on the rounding result, catching both.
- Closes #1232.

## Rationale

LLVM `fptosi` / `fptoui` return [poison](https://llvm.org/docs/LangRef.html#fptosi-to-instruction) when the source is NaN, ±inf, or outside the target integer's representable range. Poison propagates silently through downstream arithmetic and is true undefined behavior. #1192 exposed this in the implicit `float → int` coercion and the issue consolidates every FPToSI / FPToUI site in one sweep.

## Test plan

- [x] `./build/ry_tests` — 1894 tests pass (11 new `CastFloat*` death tests + INT64_MIN boundary regression)
- [x] `./build/ry test -p` — 161 Ry spec tests pass (including `math.test.ry` boundary case)
- [x] ASan + UBSan — all tests pass
- [x] TSan — all C++ tests pass
- [x] libFuzzer 60s each (`fuzz_parser` / `fuzz_json` / `fuzz_utf8`) — no crashes
- [x] Ad-hoc: `(1.0 / 0.0) as int` → `runtime error: cannot convert inf to int`, exit 1
- [x] Ad-hoc: `(-9.223372036854776e+18) as int` → `-9223372036854775808`, exit 0 (INT64_MIN accepted)
- [x] Ad-hoc: `3.14 as int` → `3`, `-3.7 as int` → `-3` (finite fast path preserved)

## Notes

- `docs/reference/types.md` gains a short "Float → Integer Runtime Checks" subsection describing the user-visible behaviour.
- `docs/reference/math.md` updates the note about `floor()` / `ceil()` / `round()` error formatting.
- `KNOWLEDGE.md` adds an entry (Tags: `codegen, conversion, runtime-check, fptosi, fptoui`) instructing future code to route through `emitCheckedFPToInt` and never call `CreateFPToSI` / `CreateFPToUI` directly.
- Emitted IR is lean: unordered comparisons (`ULT` / `UGE`) absorb the NaN case into the range check, so each call site adds 2 fcmps + 1 or + 1 condbr rather than the more obvious 3 fcmps + 2 ors.

## Summary by CodeRabbit

* **Bug Fixes**
  * Float→integer conversions (explicit casts, implicit coercions, compound ops) now perform runtime checks and terminate with exit code 1 on NaN, ±Inf, or out-of-range values.
  * Rounding functions (floor/ceil/round/trunc) now accept boundary values previously rejected and correctly handle overflow cases.
  * Runtime error text standardized to "cannot convert <value> to <type>".

* **Documentation**
  * Updated casting and math references to reflect new runtime validation and error messages.

* **Tests**
  * Added regressions validating new runtime failures and boundary acceptance.

### CODE REVIEW

File: docs/reference/math.md [hunk 1]
```diff
@@ -79,7 +79,7 @@ round(1750.0, -3)    # 2000.0
 
 Rounding uses C99 half-away-from-zero semantics (via `round(x * 10^digits) / 10^digits`), matching the one-argument form. This differs from Python's banker's rounding — for example, `round(2.675, 2) == 2.68`, not `2.67`. `NaN` and `±Inf` pass through unchanged in the two-argument forms.
 
-> **Note**: The one-argument forms (`floor(x)`, `ceil(x)`, `round(x)`) convert the result to `int`. If the argument is `Inf`, `-Inf`, or `NaN`, a runtime error occurs: `runtime error: floor()/ceil()/round() argument out of int range`.
+> **Note**: The one-argument forms (`floor(x)`, `ceil(x)`, `round(x)`) convert the result to `int`. If the rounded result is `NaN`, `±Inf`, or outside the `int` range (e.g. `ceil(9.22e+18)`, which rounds past `INT64_MAX`), a runtime error occurs: `runtime error: floor(): cannot convert <value> to int` (with the callee name prefixed).
```
[the remark this answers is not in the document]
Comment (t0k0sh1) [reply docs/reference/math.md]: Fixed in the next commit. Verified: `9.22e18` is `3.37e15` below `2^63`, so `ceil(9.22e18)` neither crosses `INT64_MAX` nor exercises the old `fabs >= 2^63` bug — the example was wrong.

Changes:
- `docs/reference/math.md`: replaced the example with `floor(1e100)` (`1e100` is clearly outside `int` range).
- `src/codegen_call.cpp`: rewrote the inline comment to state the actual motivation (uniformity with other float→int sites + the `INT64_MIN` boundary fix) instead of the misleading rounding claim.
- `changelog.d/1232-fptosi-runtime-check.md`: removed the `ceil(9.22e+18)` clause; kept only the `INT64_MIN` boundary fix, which is the actual behavioural change.

Also added a concrete error-message example to the new `Float → Integer Runtime Checks` section in `docs/reference/types.md` per the nitpick in your review summary — verified against runtime output (`(1.0 / 0.0) as int` → `runtime error: cannot convert inf to int`, etc.).

**Comment by t0k0sh1** [comment]
@coderabbitai Review-summary nitpick addressed:

**`docs/reference/types.md` (lines 434-436)** — Added a concrete error-message example to the new `Float → Integer Runtime Checks` section:

```ry
(1.0 / 0.0) as int       # runtime error: cannot convert inf to int
(0.0 / 0.0) as i32       # runtime error: cannot convert nan to i32
(-1.0) as u8             # runtime error: cannot convert -1 to u8
```

Wording verified against actual runtime output (lowercase `inf` / `nan` because the helper passes `%g` to the format string). Will be in the next commit.
